chronos/streams/asyncstream.nim

```diff
@@ -0,0 +1,908 @@
+  result.closed = 0
+  result.dump = dumpAsyncStreamReaderTracking
+  result.isLeaked = leakAsyncStreamReader
+  addTracker(AsyncStreamReaderTrackerName, result)
+
+proc setupAsyncStreamWriterTracker(): AsyncStreamTracker {.gcsafe.} =
+  result = new AsyncStreamTracker
+  result.opened = 0
+  result.closed = 0
+  result.dump = dumpAsyncStreamWriterTracking
+  result.isLeaked = leakAsyncStreamWriter
+  addTracker(AsyncStreamWriterTrackerName, result)
+
+proc readExactly*(rstream: AsyncStreamReader, pbytes: pointer,
+                  nbytes: int) {.async.} =
+  ## Read exactly ``nbytes`` bytes from read-only stream ``rstream`` and store
+  ## it to ``pbytes``.
+  ##
+  ## If EOF is received and ``nbytes`` is not yet readed, the procedure
+  ## will raise ``AsyncStreamIncompleteError``.
+  if not rstream.running():
+    raise newAsyncStreamIncorrectError("Incorrect stream state")
+
+  if isNil(rstream.rsource):
+    try:
+      await readExactly(rstream.tsource, pbytes, nbytes)
+    except TransportIncompleteError:
+      raise newAsyncStreamIncompleteError()
+    except:
+      raise newAsyncStreamReadError(getCurrentException())
+  else:
+    var index = 0
+    while true:
+      let datalen = rstream.buffer.dataLen()
+      if rstream.state == Error:
+        raise newAsyncStreamReadError(rstream.error)
+      if datalen == 0 and rstream.atEof():
+        raise newAsyncStreamIncompleteError()
+
+      if datalen >= (nbytes - index):
+        rstream.buffer.copyData(pbytes, index, nbytes - index)
+        rstream.buffer.shift(nbytes - index)
+        break
+      else:
+        rstream.buffer.copyData(pbytes, index, datalen)
+        index += datalen
+        rstream.buffer.shift(datalen)
+      await rstream.buffer.wait()
+
+proc readOnce*(rstream: AsyncStreamReader, pbytes: pointer,
+               nbytes: int): Future[int] {.async.} =
+  ## Perform one read operation on read-only stream ``rstream``.
+  ##
+  ## If internal buffer is not empty, ``nbytes`` bytes will be transferred from
+  ## internal buffer, otherwise it will wait until some bytes will be received.
+  if not rstream.running():
+    raise newAsyncStreamIncorrectError("Incorrect stream state")
+
+  if isNil(rstream.rsource):
+    try:
+      result = await readOnce(rstream.tsource, pbytes, nbytes)
+    except:
+      raise newAsyncStreamReadError(getCurrentException())
+  else:
+    while true:
+      let datalen = rstream.buffer.dataLen()
+      if rstream.state == Error:
+        raise newAsyncStreamReadError(rstream.error)
+      if datalen == 0:
+        if rstream.atEof():
+          result = 0
+          break
+        await rstream.buffer.wait()
+      else:
+        let size = min(datalen, nbytes)
+        rstream.buffer.copyData(pbytes, 0, size)
+        rstream.buffer.shift(size)
+        result = size
+        break
+
+proc readUntil*(rstream: AsyncStreamReader, pbytes: pointer, nbytes: int,
+                sep: seq[byte]): Future[int] {.async.} =
```

Comment:
I don't think all the functions has same logic. While `readUntil` and `readLine` has almost equal logic, it has different behavior for different situations. For example, when `readLine` returns empty string, `readUntil` will return an error. Also `readUntil` do not allocate new memory, `readLine` performs allocation.

Also i do not like callback approach because callback will require some state context while processing/checking/reading and so adds more complexity to the code, for example even `readUntil` and `readLine` will have different callbacks because `readLine` do not check output buffer size and just allocated a bigger string. So i dont see any benefits from using callbacks style. It will be just transfer of the logic from the function itself to some other procedure, and this callbacks will still be 70% different.